src/g_kbdnav.c

```diff
@@ -0,0 +1,2183 @@
+        {
+            if(kbdnav->kn_linetraverser->tr_inno == kbdnav->kn_ioindex)
+                break;
+        }
+    }
+    if( oc==NULL )
+    {
+        /* happens when you delete the last connection with the Delete key */
+        kbdnav->kn_state = KN_IO_SELECTED;
+    }
+    kbdnav->kn_outconnect = oc;
+}
+
+
+/* this traverses all the t_outconnections in the current outlet the user have seleced, one by one */
+void kbdnav_traverse_inlet_sources_next(t_canvas *x)
+{
+    t_kbdnav *kbdnav = canvas_get_kbdnav(x);
+
+    int n_of_connections = kbdnav_number_of_inlet_sources(x, kbdnav->kn_selobj, kbdnav->kn_ioindex);
+
+    if (kbdnav->kn_connindex+1 == n_of_connections)
+    {
+        kbdnav_traverse_inlet_sources_start(x);
+    }
+    else if(n_of_connections > 0)
+    {
+        /* search for the next connection */
+        linetraverser_start(kbdnav->kn_linetraverser, x);
+        ++kbdnav->kn_connindex;
+        int count = 0;
+        t_outconnect *oc;
+        while (  oc = linetraverser_next(kbdnav->kn_linetraverser)  )
+        {
+            if( kbdnav->kn_linetraverser->tr_ob2 == kbdnav->kn_selobj
+                && kbdnav->kn_linetraverser->tr_inno == kbdnav->kn_ioindex)
+            {
+                count++;
+                if( count == kbdnav->kn_connindex+1)
+                  break;
+            }
+        }
+        kbdnav->kn_outconnect = oc;
+    }
+    else
+    {
+        /* happens when you delete the last connection with the Delete key */
+        kbdnav->kn_state = KN_IO_SELECTED;
+    }
+}
+
+void kbdnav_traverse_inlet_sources_prev(t_canvas *x)
+{
+    t_kbdnav *kbdnav = canvas_get_kbdnav(x);
+    int n_of_connections = kbdnav_number_of_inlet_sources(x, kbdnav->kn_selobj, kbdnav->kn_ioindex);
+    int target_connection = kbdnav->kn_connindex == 0 ? n_of_connections-1 : kbdnav->kn_connindex-1;
+
+    kbdnav_traverse_inlet_sources_start(x);
+    for (int i = 0; i<target_connection; ++i){
+        kbdnav_traverse_inlet_sources_next(x);
+    }
+}
+
+void kbdnav_displayindices(t_canvas *x)
+{
+    t_gobj *gobj = x->gl_list;
+    if ( !gobj ){
+        return;
+    }
+    int i = 0;
+    int there_is_next = 1;
+    while ( there_is_next ){
+        t_object *obj = pd_checkobject(&gobj->g_pd);
+        int objindex = canvas_getindex(x,gobj);
+        sys_vgui(".x%lx.c create text %d %d -text {%d} -tags [list objindex%d obj_index_number] -fill blue\n",
+                x,
+                obj->te_xpix-10, obj->te_ypix+5,
+                i,
+                objindex);
+        //https://wiki.tcl-lang.org/page/canvas+text
+        sys_vgui(".x%lx.c create rect [.x%lx.c bbox objindex%d]  -fill white -outline blue -tag [list objindex_bkg%d obj_index_number_bkg]\n",
+                x,
+                x,
+                objindex,
+                objindex);
+        sys_vgui(".x%lx.c lower objindex_bkg%d objindex%d\n",
+                x,
+                objindex,
+                objindex);
```

Comment:
this should probably go into generic tcl/tk proc for adding tooltips.
something like

~~~tcl
proc tooltip_show {mytoplevel id posx posy text {color blue} {outline none}} {
  # ...
}
proc tooltip_hide {mytoplevel id} {
 # ...
}
~~~

(with `id` being just some identifier (probably without dots...) that allows us to identify the tooltip when we want to remove it later...)

Reply:
Agree.

Reply:
okay